### ISSUE TITLE
The semantics of the nuraft::snapshot

### ISSUE THREAD

**Issue by unixod** [eBay/NuRaft#509]
Hi,

Is it correct to assume that for each call of `read_logical_snp_obj(s, ...)` made by NuRaft, the following holds?
1.  The first parameter `s` contains an information about raft log entry at which latest snapshot was made;
2.  The value of first parameter `s`  is equal to the one which was passed to latest call of `create_snapshot` (i.e. s.get_last_log_idx() is equal to index of latest log entry included into latest snapshot);

To be more general, I'm wondering what information is passed as a first parameter to functions:
- `create_snapshot(snapshot&, ...)`
- `read_logical_snp_obj(snapshot&, ...)`
- `save_logical_snp_obj(snapshot&, ...)`
- `apply_snapshot(snapshot&, ...)`

**Comment by greensky00** [comment]
Hi @unixod 

1. `s` contains the information of the snapshot that the current NuRaft instance (leader) is transmitting to its follower. It can be the latest snapshot, or older than that.
2. Same as above, it can be older than the latest snapshot. More precisely, it is the snapshot that was the latest one at the moment the leader began the snapshot transmission. More snapshots can be created since then, but `s` will remain unchanged unless a new snapshot transmission begins.

In short,
* `create_snapshot`: `s` contains the latest snapshot info, that the state machine has to create.
* `read_logical_snp_obj`: `s` contains the snapshot that the current leader is transmitting. `s` will remain the same until the transmission either succeeds or fails. If it fails, `read_logcial_snp_obj` will be retried with newer `s`.
* `save_logical_snp_obj`, `apply_snapshot`: same as `read_logical_snp_obj`.

**Comment by grigoryan-sergey** [comment]
Hi @greensky00 

Thanks, the answer clarified a lot of things. But I still have several questions about.

1. Looking at the calculator example I observed that the state machine stores only last 3 snapshots.  How was this number determined? Is it safe to store any number of snapshots, or there is a limit for it? Will everything work correctly, if we store just one snapshot for each node?
2. Looking at `read_logical_snp_obj` function. There is an if block [there](https://github.com/eBay/NuRaft/blob/32502e410a87a9e818728ff6696eda83760bc06b/examples/calculator/calc_state_machine.hxx#L123C1-L123C45), and I am wondering what are the concrete cases that we can enter that block?
3. Also why do we return `0` in that block if that is considered to be failure of snapshot reading (I mean why don't we return a negative value)?
4. Where does NuRaft take that first `snapshot` argument from? Am I correct that it gets it from calling `last_snapshot` on the leader node?
5. Because the reference of the first  `snapshot` argument is non const in all of the functions above, I suppose that user can modify it. Am I correct and if yes, what are those cases when the first argument can be modified?

**Comment by unixod** [comment]
Thanks!
